### ISSUE TITLE
FP S3516 InvariantReturnCheck

### ISSUE THREAD

**Issue by vilchik-elena** [SonarSource/SonarJS#609]
RSPEC-3516
```javascript
function foo() {
  var a = foobar - 1;
  var d;

  if (cond1) {
    if (cond2) {
      d = a;
    } else {
      d = a - 1;
    }
  } else {
    if (cond3) {
      return a;
    }
    if (cond4) {
      d = a - 1;
    } else {
      d = a;
    }
  }
  return d; // this line got executed by DFA only once since program states are considered equal (in all cases 'd' is number)
}
```

**Comment by MarekMazu** [comment]
Another example of false positive:


```javascript
function sonarFalsePositive(condition) {
     let foo = 1;
     if(condition) return foo;
     while(foo < 5) {        
        foo++;                   //also ++foo is giving false positive, but not foo += 1
     }
     return foo;
}

console.log(sonarFalsePositive(true));
console.log(sonarFalsePositive(false));
```

**Comment by charlesritchea** [comment]
Here's another one where a collection is modified
```
const latestMarker = new AnnotationCollection<AnnotationModel>();
if (!this._markerCollection) {
    // return empty makers
    return latestMarker;
}
const markers = this._markerCollection.toArray();
const data = uniqWith(markers, (arrVal: AnnotationModel, othVal: AnnotationModel) => {
    const {geometry: geo1} = arrVal;
    const {geometry: geo2} = othVal;
    return isEqual(geo1.coordinates, geo2.coordinates);
});
data.forEach(item => latestMarker.add(item));
return latestMarker;

```

**Comment by vilchik-elena** [comment]
Closing this ticket as this rule does not use data flow analysis anymore
